### ISSUE TITLE
Print jobs get canceled at printer

### ISSUE THREAD

**Issue by mcp292** [OpenPrinting/cups#1092]
# Purpose

This issue is being made at the request of @zdohnal in #1090.

# Have

I have a Brother HL-L2325DW. I set it up as a driverless network printer through the GNOME settings using add printer and leaving defaults.

Running Fedora 40 (Workspace Edition) with GNOME on a Thinkpad T14s laptop. OpenPrinting CUPS 2.4.11.

# Problem 

Print jobs get canceled at printer repeatedly. Usually the first print after powering on the printer goes through, and anything else gets canceled.

# Tried

I've tried  to remove the printer and re-add, change settings, print from different places (lp, lpr, firefox, document viewer), factory reset printer, reset printer network settings.

# Logs

* [multiple_prints.txt](https://github.com/user-attachments/files/17561910/multiple_prints.txt) - Here is a session where the first document sent printed successfully, and the rest were canceled at printer. Search for `Unable to add document to print job.` which appears in red in my terminal. Sorry I cannot preserve the color coding. There's some odd stuff about "cropping" near the end. These were set to print from Firefox.
* [ppd_before_printing.txt](https://github.com/user-attachments/files/17562084/ppd_before_printing.txt) - Added `DirtyCleanInterval 0` to `cupsd.conf`, issued `sudo systemctl restart cups`, removed printer, re-added printer, copied `/etc/cups/ppd`. (Removed and added printer through GNOME settings.)
* [ppd_after_printing.txt](https://github.com/user-attachments/files/17562057/ppd_after_printing.txt) - Printed document through Firefox. Got a GNOME notification saying printing complete but nothing happened. Nothing printed.

**Comment by mcp292** [comment]
* [multiple_prints.txt](https://github.com/user-attachments/files/17561910/multiple_prints.txt) - Here is a session where the first document sent printed successfully, and the rest were canceled at printer. Search for `Unable to add document to print job.` which appears in red in my terminal. Sorry I cannot preserve the color coding. There's some odd stuff about "cropping" near the end. These were set to print from Firefox.

**Comment by zdohnal** [comment]
Ok, ```*cupsMandatory``` is defined as ```*cupsMandatory: "value1 ... valueN", but we don't use quotes there.

Can you try editing the PPD and reinstalling printer with the edited PPD?

Like you change in the PPD (/etc/cups/ppd/HL-L2325DW.ppd) the line:

```
*cupsMandatory: attributes-charset attributes-natural-language printer-uri
```

to 

```
*cupsMandatory: "attributes-charset attributes-natural-language printer-uri"
```

and then reinstall the printer this way:

```
$ sudo lpadmin -p HL-L2325DW -P /etc/cups/ppd/HL-L2325DW.ppd -E
```

and try to print?

Ad the problem itself - does it happen even for the same file? First print goes well, but the next of the same file does not print? (from the same application)

Additionally, can you check if you get the same situation if you print the same file from command line? The command is:

```
lp -d HL-L2325DW <file>
```

**Comment by mcp292** [comment]
Okay, I changed the line in the ppd, ran the install line, then tried to print with two-sided, 9-up (what I used when producing the logs). Printing canceled at printer.

I tried again with two-sided, 1-up and it went through. This and above with Firefox.

Tried to print two-sided, 9-up for consistency with:

```sh
lp -d HL-L2325DW -o number-up=9 -o sides=two-sided-long-edge  ~/downloads/printing.pdf
```

Canceled at printer.

Tried two-sided, 1-up (what worked in FF)f;

```sh
lp -d HL-L2325DW-2 -o sides=two-sided-long-edge ~/downloads/printing.pdf
```

Canceled at printer.

Tried the above line again to try to rule out "luck of the draw". Canceled at printer.

**Comment by mcp292** [comment]
Did you need me to explicitly remove the printer before running the `lpadmin` install line?

**Comment by zdohnal** [comment]
What is "HL-L2325DW"? The reinstalled printer with the updated PPD?

There is no need to restart the service or remove the original printer - cupsd should update the cache and files accordingly asap. However, cupsd might rewrite the PPD to incorrect form again, but since even the first printing was cancelled, it might be red herring.

So to simplify this and rule out any possible divergences, stick with CUPS utils and no options from now (so no Gnome, no Firefox, no additional options, until we found out what triggers it):

1. check if the issue happens with IPP Everywhere - you install it by:

```
$ lpadmin -p <eve_name> -v ipp://<ip or hostname>/ipp/print -m everywhere -E
```

and print to it twice the same file (choose of course some one page PDF, so you don't get swarmed by papers and we can rule out this possibility):

```
$ lp -d <eve_name> <file>.pdf
```

If those two print okay, try larger file and add options with lp - if that prints ok, try printing to the queue from Firefox or other app you use.

If all this work, we can say the issue is related to driverless driver and not to IPP Everywhere. If it does not, then the issue might be in cups in general or in cups-filters, or in printer itself - so type of no-driver solution does not matter.

2. If IPP Everywhere worked, let's focus on "driverless", which is the driver Gnome sets up for the printer. You can reinstall it by this way:

```
$ lpadmin -p <name> -v ipp://<ip or hostname>/ipp/print -m driverless:ipp://<ip or hostname>/ipp/print -E
```

then try print twice with lp again, first one page PDF without options, then multipage pdf and options, and in the end via Firefox.

I'm sorry for lot of printing, but there can be different things in play, so I need to find out the minimal reproducer which triggers the issue - especially the fact it shows only after first printing is tricky.

**Comment by mcp292** [comment]
> I'm sorry for lot of printing, but there can be different things in play, so I need to find out the minimal reproducer which triggers the issue - especially the fact it shows only after first printing is tricky.

No worries about the printing. This problem costs me a considerable amount of time and focus so fixing it is well worth the additional time and resources expended. I appreciate your help.

> What is "HL-L2325DW"? The reinstalled printer with the updated PPD?

Regarding HL-L2325DW vs HL-L2325DW-2. That is a very good catch. Sorry for the confusion. From what I recall, there is only HL-L2325DW-2 and I was listing it here without the two to avoid added complexity. I understand that in doing so I made things more complicated. I will not modify anything from here on out. 

To answer your question, both are the reinstalled printer with updated PPD. Reinstalled as instructed without removing the original.

> So to simplify this and rule out any possible divergences, stick with CUPS utils and no options from now (so no Gnome, no Firefox, no additional options, until we found out what triggers it):

Perfect, got it!

I'm going to list what I do so you can validate each step.

> 1. Check if the issue happens with IPP Everywhere 

Finding URI:

```sh
$ lpinfo -v
file cups-brf:/
network beh
network https
network http
network ipps
network ipp
direct hp
network socket
network lpd
network smb
direct hpfax
network dnssd://Brother%20HL-L2325DW._ipp._tcp.local/?uuid=e3248000-80ce-11db-8000-00410ef027f1
network lpd://BRW00410EF027F1/BINARY_P1
network ipp://Brother%20HL-L2325DW._ipp._tcp.local/
```

Installing:

```sh
$ lpadmin -p ipp_everywhere -v ipp://Brother%20HL-L2325DW._ipp._tcp.local/ipp/print -m everywhere -E
```

Printing twice:

```sh
$ lp -d ipp_everywhere printing.pdf
request id is ipp_everywhere-647 (1 file(s))
```

Success!

```sh
$ lp -d ipp_everywhere printing.pdf
request id is ipp_everywhere-648 (1 file(s))
```

Success!

Printing large file with options:

```sh
$ lp -d ipp_everywhere -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf 
request id is ipp_everywhere-649 (1 file(s))
```

Success!

Trying same file ([webpage](https://en.wikipedia.org/wiki/Printing)) from Firefox with same options:

Success!

Trying same downloaded file from PDF Viewer with same options:

Canceled at printer.

Tried from commandline again to see if any settings were overwritten by PDF Viewer:

```sh
$ lp -d ipp_everywhere -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf 
request id is ipp_everywhere-654 (1 file(s))
```

Success

Tried from PDF Viewer again while checking journal:

[pdf_log.txt](https://github.com/user-attachments/files/17611607/pdf_log.txt)

(The evince line is not part of the log, but instead something that gets printed to stdout by the PDF Viewer. It printed out while logs were printing.)

> If all this work, we can say the issue is related to driverless driver and not to IPP Everywhere. If it does not, then the issue might be in cups in general or in cups-filters, or in printer itself - so type of no-driver solution does not matter.
> 
> 2. If IPP Everywhere worked, let's focus on "driverless", which is the driver Gnome sets up for the printer. 

Since the only thing that failed is PDF Viewer, I will move forward and test driverless.

Re-installing:

```sh
$ lpadmin -p driverless -v ipp://Brother%20HL-L2325DW._ipp._tcp.local/ipp/print -m driverless:ipp://Brother%20HL-L2325DW._ipp._tcp.local/ipp/print -E
lpadmin: Unable to open PPD "/tmp/71d2967280f8c": Missing PPD-Adobe-4.x header on line 0.
```

Can't proceed to test printing. Awaiting instruction.

**Comment by mcp292** [comment]
I highly suspect this has something to do with print scaling. Scaling options are not present in the GUI system print when using an IPP Everywhere printer. But are for driverless. This alone is not enough evidence.

**Comment by mcp292** [comment]
Let me know if you need me to try anything else.

**Comment by zdohnal** [comment]
@mcp292 thank you very much for the detailed testing! And I'm sorry for delay, I was stuck with other things :(

So I guess we can rule out the fact the PPD change would make a difference, since repeated printing brought the same results, and the issue is probably in filter chain, since both models fail when printing from PDF viewer, but driverless model seems to be more prone to the cancellation.

I wonder about print-scaling - it is a good find that the setting is not present in GUI of PDF viewer, but I don't see the setting in the previous logs from driverless either (there is only the default one applied in the filter, which happens for both prints, successful and failed).

So based on your findings, let's see the logs where those two models started to differ:

```
lp -d ipp_everywhere -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf 
``` 

and

```
lp -d driverless -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf 
```

Please provide me logs for both jobs, and both PPDs. If there is any change in behavior in comparison to the past, do let me know - I count on it will behave the same - everywhere prints, driverless gets canceled.

Hopefully it will give us some ideas which will lead even to fixing PDF viewer printing.

**Comment by mcp292** [comment]
> I'm sorry for delay...

No apology needed. I appreciate your help.

## ipp_everywhere

```sh
lp -d ipp_everywhere -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
```

Printing succeeded.

[ipp_everywhere_log.txt](https://github.com/user-attachments/files/17834747/ipp_everywhere_log.txt)

[ipp_everywhere_ppd.txt](https://github.com/user-attachments/files/17834755/ipp_everywhere_ppd.txt)

## driverless

```sh
lp -d driverless -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
```

Printing canceled.

[driverless_log.txt](https://github.com/user-attachments/files/17834758/driverless_log.txt)

[driverless_ppd.txt](https://github.com/user-attachments/files/17834762/driverless_ppd.txt)

**Comment by mcp292** [comment]
Bumping, in case it fell off your radar. Still having issues daily.

**Comment by zdohnal** [comment]
@mcp292 I'm sorry, I was focused on my work and just tried to answer on the new issues, so I missed the backlog :( .

I've checked the logs now and it looks to be connected to print quality.

IPP Everywhere:

```
*DefaultResolution: 300dpi
*OpenUI *cupsPrintQuality: PickOne
*OrderDependency: 10 AnySetup *cupsPrintQuality
*en_US.Translation cupsPrintQuality/Print Quality: ""
*DefaultcupsPrintQuality: Normal
*cupsPrintQuality Draft: "<</HWResolution[300 150]>>setpagedevice"
*en_US.cupsPrintQuality Draft/Draft: ""
*cupsPrintQuality Normal: "<</HWResolution[300 300]>>setpagedevice"
*en_US.cupsPrintQuality Normal/Normal: ""
*cupsPrintQuality High: "<</HWResolution[1200 1200]>>setpagedevice"
*en_US.cupsPrintQuality High/High: ""
*CloseUI: *cupsPrintQuality
```

driverless
```
*DefaultResolution: 600dpi
*OpenUI *cupsPrintQuality/Print Quality: PickOne
*OrderDependency: 10 AnySetup *cupsPrintQuality
*DefaultcupsPrintQuality: Normal
*cupsPrintQuality Draft/Draft: "<</HWResolution[300 300]>>setpagedevice"
*cupsPrintQuality Normal/Normal: "<</HWResolution[600 600]>>setpagedevice"
*cupsPrintQuality High/High: "<</HWResolution[1200 1200]>>setpagedevice"
*CloseUI: *cupsPrintQuality
```

Only the highest resolution matches, draft and normal have a different numbers and  "normal" is used by default.

So it looks like the printer somehow manages to process smaller files with higher resolution (like it influences the Apple raster file generation at the end of filtering), but fails for larger ones if the job is set as resolution 600x600.

Can you get attrs.txt file with output from ipptool command:

```
$ ipptool --ippserver attrs.txt ipp://BRW00410EF027F1.local:631/ipp/print get-printer-attributes.test
```

it will get us IPP attributes from the printer, so we will see what resolutions it reports so we can track what driverless PPD generator did wrong, or if there is discrepancy on the printer side.

You can try taking driverless PPD, change the normal quality resolution, create a new queue with it (the same connection as before) and see if the testing command passes.

In general you:

```
$ sudo cp /etc/cups/ppd/driverless.ppd /tmp
change "<</HWResolution[600 600]" to "<</HWResolution[300 300]" in new PPD
$ sudo lpadmin  -p new-driverless -v ipp://BRW00410EF027F1.local:631/ipp/print -P /tmp/driverless.ppd -E
$ lp -d new-driverless -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
```

If the suspicion is correct, new-driverless should now print.

**Comment by mcp292** [comment]
Nice Digging!

== Get IPP Attributes

```sh
$ ipptool --ippserver attrs.txt ipp://Brother%20HL-L2325DW._ipp._tcp.local/print get-printer-attributes.test
```

[attrs.txt](https://github.com/user-attachments/files/18099352/attrs.txt)

== Install New Printer with Edited PPD

=== Edit PPD

[new_driverless_ppd.txt](https://github.com/user-attachments/files/18099354/new_driverless_ppd.txt)

=== Install

```sh
$ lpadmin -p new_driverless -v ipp://Brother%20HL-L2325DW._ipp._tcp.local/ipp/print -P /etc/cups/ppd/new_driverless.ppd -E
lpadmin: Printer drivers are deprecated and will stop working in a future version of CUPS.
```

Interestingly, `lpinfo -v` didn't find the ipp URI until I rebooted the printer. Maybe when it encounters the rastering error it drops off? I'm guessing that my last print left my printer in a non-working state, such that I would've had to restart to print the next document successfully. Recording this in case it helps you later.

=== Test

```sh
$ lp -d new_driverless -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
request id is new_driverless-895 (1 file(s))
```

Success!

So then I tried driverless as a sanity check:
```sh
$ lp -d driverless -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
request id is driverless-896 (1 file(s))
```

Success!

And then ipp_everwhere:
```sh
$ lp -d ipp_everywhere -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
request id is ipp_everywhere-897 (1 file(s))
```

Canceled.

Since driverless worked when it previously didn't, I shut off printer and tried with driverless first:
```sh
$ lp -d driverless -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
request id is driverless-898 (1 file(s))
```

Success!

new_driverless:
```sh
$ lp -d new_driverless -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
request id is new_driverless-899 (1 file(s))
```

Success!

ipp_everywhere:
```sh
$ lp -d ipp_everywhere -o number-up=9 -o sides=two-sided-long-edge printing_full.pdf
request id is ipp_everywhere-900 (1 file(s))
```

Success!

Very odd. Now they all printed. So then I rebooted printer again, tried new_driverless, driverless, ipp_everwhere in that order (again), and now ddriverless canceled and the other two printed.

Not sure what's going on but if I had to guess, each print is filling up an error log and at a certain point the printer decides not to proceed. Meaning, maybe even the successful prints are throwing off warnings, until a threshold is hit where no more jobs are taken, or where the error causing the warning, causes some bits to flip that it shouldn't.

I'm realizing now that there probably isn't much you can do without a log, so started recording a log, shutoff my printer, turned it back on, and printed with new_driverless, driverless, ipp_everwhere, in that order. They all printed. So then I ran that sequence again, without rebooting, and new_driverless printed and the rest failed.

[2024-12-11_log_1.txt](https://github.com/user-attachments/files/18099359/2024-12-11_log_1.txt)

[2024-12-11_log_2.txt](https://github.com/user-attachments/files/18099361/2024-12-11_log_2.txt)

Let me know how to proceed.

**Comment by mcp292** [comment]
Bump.

**Comment by mcp292** [comment]
I do think you're onto something with the resolution. I did a big batch print (which never works) where I put all normal resolutions first and high resolution last and sent them back-to-back to the driverless queue and all three normals printed but the last high res failed.

**Comment by mcp292** [comment]
Bump.

**Comment by zdohnal** [comment]
Hi,

I'm sorry, I've checked the PPDs, IPP attributes and logs and looks like actually driverless is the one generating print-quality+resolution correctly from urf-supported, while CUPS IPP Everywhere uses the lowest DPI for some reason for normal quality, and divides yres for draft resolution. So unless ghostscript sometimes generates the raster differently, the print quality settings is correct for driverless.

To be honest, I'm kind of lost by the randomness you've found ... from the logs I see the jobs are mixed together (the previous is still running, while the newer is already created at printer), so maybe the printer sometimes does not handle multiple jobs at the same time, resulting into killing any jobs which can overflow the device memory?

Can you try waiting with the new job until the previous one is done (there is `Job completed` in cupsd logs for the specific job)?

Otherwise we can send the filtering output into a file, and then try to pass the file directly to printer - we could catch if there is a difference between outputs of different jobs, and check if we send the same file several times, whether we get the same result (the same broken result would mean problem in the generated file, if the result is random then probably the issue is in firmware...)

**Comment by zdohnal** [comment]
You can do the printing to a file via ippeveprinter:

```
$ ippeveprinter -a <attr file you attached> -D file:<full path to a directory> "<name>" &
```

The app should simulate your printer and it will be seen in printer list under "<name>" - printouts will be in /chosen_dir/ippeveprinter.clientid/ dir as file `<jobid>-<jobname>.<document-format-suffix>. You can print to it via classic `lp`, which options you used.

After this printing finishes, take the printout file and send it to the print queue for real printer (like `ipp_everywhere`) with options - IIUC cupsd should recognize the file is already ready to be printed and only encapsulates options for the printer, which are supposed to be applied on the printer itself, not during filtering.

**Comment by jsmeix** [comment]
A side note FYI regarding
```
printer sometimes does not handle multiple jobs at the same time
```

I remember it from the past (long time ago when I had tested it)
that a well known standard network printer for business usage
from a reputable established printer manufacturer
badly failed when I sent multiple jobs at the same time.
It hung up and became unresponsive to that the only way out
for me in practice was a hard printer reset via power off/on.
I think at that time I had tested only via plain TCP socket.
Perhaps I had also tested LPD but I don't remember details.
Certainly I had not tested IPP.

Cf.
https://en.opensuse.org/SDB:Printing_via_TCP/IP_network#Network_printer_or_printserver_box_does_not_work_reliably
and
https://en.opensuse.org/SDB:Using_Your_Own_Backends_to_Print_with_CUPS#A_hardcoded_backend_for_a_network_printer
therein in particular the part about an "optimistic backend for a network printer".

**Comment by mcp292** [comment]
That's a very interesting deduction. Since I have experienced these issues both with batch prints, and spaced out prints that wait for completion, I'm going to rule out that possibility for now. We can always revisit it.

I like where you're heading with the filtering, but what stands out to me most is the firmware possibility. My firmware is outdated by many versions (1.72 -> 1.77). I am on linux and haven't been able to update it. I recently succeeded and will report more on that front to help other users. (I'll put something together and link it here.) But for now, the new firmware seems to be working well. I need more time to really test it, so I'll let you know if that's the fix.
